firedrake/assemble.py

```diff
@@ -186,13 +187,19 @@ def _assemble(f, tensor=None, bcs=None, form_compiler_parameters=None,
         form_compiler_parameters = {}
     form_compiler_parameters["assemble_inverse"] = inverse
 
+    topology = f.ufl_domains()[0].topology
     for m in f.ufl_domains():
         # Ensure mesh is "initialised" (could have got here without
         # building a functionspace (e.g. if integrating a constant)).
         m.init()
-        if m.topology != f.ufl_domains()[0].topology:
+        if m.topology != topology:
             raise NotImplementedError("All integration domains must share a mesh topology.")
 
+    for o in chain(f.arguments(), f.coefficients()):
```

Comment:
I think both `f.arguments()` and `f.coefficients()` return a tuple, so you could just write `f.arguments() + f.coefficients()`.